File: backend/policy_reviews.py
"""Project policy dates from their existing authoritative Review obligations."""
async def sync(db, review):
    policy_id = review.get("policy_id")
    if not policy_id:
        return
    scope = {"policy_id": policy_id, "client_id": review["client_id"]}
    reviews = await db.reviews.find(scope, {"_id": 0}).to_list(None)
    due = sorted(r["due_date"] for r in reviews if r.get("due_date") and r.get("status") not in ("completed", "cancelled"))
    completed = [o["completed_at"] for r in reviews for o in r.get("occurrences", []) if o.get("completed_at")]
    completed += [r["completion_date"] for r in reviews if r.get("status") == "completed" and r.get("completion_date")]
    fields = {"next_review_date": due[0] if due else None, "schedule_from_reviews": True}
    if completed:
        fields["last_reviewed_at"] = max(completed)
    await db.policies.update_one(scope, {"$set": fields})

File: frontend/src/preview/policyReviews.js
// Keep document lifecycle separate from the existing recurring Review lifecycle.
export function syncPolicyReview(db, review) {
  if (!review.policy_id) return;
  const policy = db.policies.find(p => p.policy_id === review.policy_id && p.client_id === review.client_id);
  if (!policy) return;
  const reviews = db.reviews.filter(r => r.policy_id === policy.policy_id && r.client_id === policy.client_id);
  const due = reviews.filter(r => r.due_date && !['completed', 'cancelled'].includes(r.status)).map(r => r.due_date).sort();
  const completed = reviews.flatMap(r => [
    ...(r.occurrences || []).map(o => o.completed_at),
    ...(r.status === 'completed' ? [r.completion_date] : [])
  ]).filter(Boolean).sort();
  Object.assign(policy, {next_review_date: due[0] || null, schedule_from_reviews: true});
  if (completed.length) policy.last_reviewed_at = completed[completed.length - 1];
}
